app/controllers/application_controller.rb

```diff
@@ -53,57 +33,30 @@ def require_admin
     handle_error(message)
   end
 
-  def html_request?
-    request.format.symbol == :html
-  end
-
-  def api_request?
-    [:json].include? request.format.symbol
-  end
-
-  def api_key
-    request.headers['HTTP_API_KEY']
-  end
-
-  def github_id_from_api_key(api_key)
-    return nil if api_key.nil?
+  private
 
-    client = Octokit::Client.new(access_token: api_key)
-    client.user.id
-  rescue Octokit::Unauthorized => e
-    raise UnauthorizedError::GitHubAuthenticationError, "Error authenticating via GitHub: #{e.message}"
-  end
+  def authenticator
+    if @authenticator.nil?
+      @authenticator = if html_request?
+                         WebAuthenticator.new(self)
+                       elsif api_request?
+                         ApiAuthenticator.new(self)
+                       else
+                         fail UnauthorizedError, "Unable to authenticate"
+                       end
+    end
 
```

Comment:
This is cool, what I was talking about with the one true place where the questions about request types were made. For that reason I would get rid of the #html_request? and #api_request? methods. Otherwise that exposure will encourage new code to use those methods instead of the authenticators you have setup.